patch sequence page (#1205)
ocaml/v3.ocaml.org: data/tutorials/gs_01_a_first_hour_with_ocaml.md

---
id: first-hour
title: Your First Day with OCaml
description: >
  Discover the OCaml programming language in this longer tutorial that takes you
  from absolute beginner to someone who is able to write programs in OCaml.
category: "getting-started"
date: 2021-08-06T17:11:00-00:00
---

# Your First Day with OCaml

This tutorial will not only give you the basics of OCaml, but it will
also guide you through creating your first applications with OCaml.
You can follow along with a basic OCaml installation,
as described in [Up and Running](/docs/up-and-running).

Alternatively, you may follow almost all of it by running OCaml in your browser
using the [OCaml Playground](https://ocaml.org/play), with no installation required.

On macOS/iOS/iPadOS, you can download this [all-in-one package on the App Store](https://apps.apple.com/app/ocaml-learn-code/id1547506826). It contains an editor side-by-side with an interactive toplevel. Plus, it's free and [open source](https://github.com/GroupeMINASTE/OCaml-iOS).

**Note:** The OCaml code examples in this tutorial start with a prompt `#`.

> This prompt is not part of the code, it's a hint to you that the code

Reviewer: ```suggestion
This prompt is not part of the code; it's a hint to you that the code
```

> following it should be either entered into an interactive toplevel, or

Reviewer: ```suggestion
following it should be either entered into an interactive toplevel or
```

into the OCaml Playground.
## Running OCaml Programs

> ### Using an OCaml toplevel (REPL)

Reviewer: ```suggestion
### Using an OCaml Toplevel (REPL)
```

We recommend using UTop (see [here](/docs/up-and-running#using-the-ocaml-toplevel-with-utop)),

> but alternatively, the basic toplevel of OCaml (`ocaml` command) can also be used. For installation instructions,

Reviewer: ```suggestion
but alternatively, the basic OCaml toplevel (`ocaml` command) can also be used. For installation instructions,
```

```

Note that `#use` is not part of the OCaml language proper; it's an instruction
to the OCaml toplevel only. Also, for `ocaml` or `utop` to find your file, you
may need to start `ocaml` or `utop` inside the directory of your file, use a
relative path that takes into account the directory you started the toplevel in,
or use an absolute path.

### Using the OCaml Playground

The in-browser [OCaml Playground](https://ocaml.org/play), which you can
use without installing anything, has this interface:

![OCaml Playground](/media/tutorials/ocaml-playground.png)

> There are two windows: the editor window (on the left) is where you write code,

Reviewer: ```suggestion
There are two windows: the editor window (on the left) is where you write code, and
```

- : int = 3
```

> This defines two variables `a` and `b` with the values of `1` and `2` respectively,

Reviewer: ```suggestion
This defines two variables `a` and `b` with the values of `1` and `2`, respectively,
```

and then uses them in the expression `a + b`, resulting in the value of `3`.
It's important to note that this is just a single expression.
The OCaml toplevel knows the expression isn't complete until it sees the `;;`.

## Functions

The `let` keyword can also be used to define a function:

```ocaml
# let square x = x * x;;
val square : int -> int = <fun>

# square 50;;
- : int = 2500
```

This defines a function named `square` which has one argument, namely `x`, and
its result is equal to (`=`) the result of the expression `x * x`. The expression
that defines the result of a function is called *function body*.

When we apply the function as `square 50;;`, this evaluates to the function body
of `square` where `x` is bound to `50`, and thus, we get a result of `50 * 50 = 2500`.

> When using `let` to define a function, the first identifier is the function name (`square`, above),
> then any additional identifiers are the different arguments of the function. In our example above,
> the `square` function has only one argument `x`.

Reviewer: ```suggestion
When using `let` to define a function, the first identifier is the function name (`square`, above).
Any additional identifiers are the different arguments of the function. In our example above,
the `square` function has only one argument `x`.
```

- : int list = [1; 2; 3]
```

> Non-empty lists have a head (its first element) and a tail (the list of the rest
> of its elements):

Reviewer: ```suggestion
Non-empty lists have a head (its first element) and a tail (the list of its remaining elements):
```

(The syntax `fun` ... `->` ... is used to build a function without a name, which
we'll only use in one place in the program.)

We can apply a function by providing only some of its arguments. This is called *partial

> application* and the value we get back from such a partial application is a new function. For example:

Reviewer: ```suggestion
application*, and the value we get back from such a partial application is a new function. For example:
```

## Other Built-In Types

> We have seen basic data types like `int`, and our first compound data type, the

Reviewer: ```suggestion
We have seen basic data types like `int` and our first compound data type, the
```

list. Let us look at two more data types of interest.

First we have tuples, which are fixed-length collections of elements of any type:

```ocaml
# let t = (1, "one", '1');;
val t : int * string * char = (1, "one", '1')
```

> Notice how the type is written: the individual types that make up a tuple
> are separated by `*`.

Reviewer: ```suggestion
Notice how the type is written: `*` separates the individual types that make up a tuple.
```
